### PR TITLE
feat(upload): support itemRender

### PR DESCRIPTION
**Pull request by kerm1it**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

close [#16047](https://github.com/ant-design/ant-design/issues/16047)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Upload support custom item of uploadList with `itemRender`. |
| 🇨🇳 Chinese | 上传组件添加 `itemRender` 以支持自定义上传列表项。        |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

-----
[View rendered components/upload/demo/drag-sorting.md](https://github.com/ant-design/ant-design/blob/feat-16047/components/upload/demo/drag-sorting.md)
[View rendered components/upload/index.en-US.md](https://github.com/ant-design/ant-design/blob/feat-16047/components/upload/index.en-US.md)
[View rendered components/upload/index.zh-CN.md](https://github.com/ant-design/ant-design/blob/feat-16047/components/upload/index.zh-CN.md)

### CODE REVIEW

**Review by xrkffgg — APPROVED** [review]

**Comment by jayce712** [comment]
when will this pr merge and publish? I need this feature...😃

**Comment by Kermit-Xuan** [comment]
Maybe this will be published in `antd@4.6.2`.

**Comment by zombieJ** [comment]
`4.6.0` 提前发了，这个要放到 `4.7.0`  里发。

**Comment by zombieJ** [comment]
拖拽样式可以再平滑一些，划线时尽量不要抖动。另外最后一个拖拽会展示不正确：
Kapture 2020-08-25 at 10 09 10

**Comment by zombieJ** [comment]
Conflict

**Comment by afc163** [comment]
Conflict

**Comment by Kermit-Xuan** [comment]
冲突解决了。

@zombieJ 

- 拖拽的时候设置了上下边框占位，防止划线时抖动；
- 最后一个拖拽是因为使用了 Tooltip，会导致拖拽的时候样式错乱，我单独定义了一个 `errorNode`，把 container 放在了 `body` 中，现在已经好了。

File: components/upload/index.en-US.md [hunk 1]
```diff
@@ -45,6 +45,7 @@ Uploading is the process of publishing information (web pages, text, pictures, v
 | transformFile   | Customize transform file before request | Function(file): string \| Blob \| File \| Promise&lt;string \| Blob \| File> | - |  |
 | iconRender | Custom show icon | (file: UploadFile, listType?: UploadListType) => ReactNode | - |  |
 | progress | Custom progress bar | [ProgressProps](/components/progress/#API) (support `type="line"` only) | { strokeWidth: 2, showInfo: false } | 4.3.0 |
+| itemRender | Custom item of uploadList | (originNode: React.ReactElement, file: UploadFile, fileList?: object\[]) => React.ReactNode | - | 4.7.0 |
```
Comment (zombieJ) [inline components/upload/index.en-US.md:48]: `React.ReactElement` => `ReactElement`

**Review by zombieJ — APPROVED** [review]

**Comment by afc163** [comment]
@Kermit-Xuan 可以给 https://github.com/ant-design/ant-design/issues/8387 搞个例子了。

**Comment by Kermit-Xuan** [comment]
我写了例子，可以预览看看。

**Comment by afc163** [comment]
哦，看到了。

**Comment by galbimandu** [comment]
Is there a codesandbox example?

**Comment by clkz** [comment]
拖拽排序的官方实例代码还是有点问题，位置交换后元素对应的data.index并没有更新，后面通过fileList.findIndex(id)来解决了.
```
const [{ isOver, dropClassName }, drop] = useDrop(
    () => ({
      accept: type,
      collect: monitor => {
        const { index: dragIndex } = monitor.getItem() || {};
        if (dragIndex === index) {
          return {};
        }
        return {
          isOver: monitor.isOver(),
          dropClassName: dragIndex < index ? ' drop-over-downward' : ' drop-over-upward',
        };
      },
      drop: item => {
        moveRow(item.index, index);
      },
    }),
    [index],
  );
```
代码中：```const { index: dragIndex } = monitor.getItem() || {}; ``` 这一句的index还是排序前的index。
目前是换成了这样：
```  
const { id } = monitor.getItem() || {};
const dragIndex = fileList.findIndex(item => item.uid === id);

... 

const [, drag] = useDrag(
    () => ({
      item: { type: 'DragableItem', index, id: file.uid },
      collect: (monitor) => ({
        isDragging: monitor.isDragging()
      })
    }),
    []
 );
```

**Comment by luotongzhou** [comment]
还是没办法对Uploadlist做自定义呀，item我不想改动，但我想在外层套个盒子

**Comment by renerlemes** [comment]
@kerm1it how to use this?

**Pull request merged**